File: uri/estructuras/Fibos Sequence/1701.cpp
#include <cstdio>
#include <cstring>

typedef long long ll;
const ll MOD = 1000000007;

struct Mat2 {
    ll a[2][2];
    Mat2() { memset(a, 0, sizeof(a)); }
};

Mat2 mul2(const Mat2& A, const Mat2& B) {
    Mat2 C;
    for (int i = 0; i < 2; i++)
        for (int j = 0; j < 2; j++)
            for (int k = 0; k < 2; k++)
                C.a[i][j] = (C.a[i][j] + A.a[i][k] * B.a[k][j]) % MOD;
    return C;
}

Mat2 pow2(Mat2 M, ll n) {
    Mat2 R;
    R.a[0][0] = R.a[1][1] = 1;
    while (n > 0) {
        if (n & 1) R = mul2(R, M);
        M = mul2(M, M);
        n >>= 1;
    }
    return R;
}

void fib(ll n, ll &fn, ll &fn1) {
    if (n == 0) { fn = 0; fn1 = 1; return; }
    Mat2 base;
    base.a[0][0] = base.a[0][1] = base.a[1][0] = 1;
    Mat2 R = pow2(base, n);
    fn = R.a[1][0];
    fn1 = R.a[0][0];
}

struct Mat5 {
    ll a[5][5];
    Mat5() { memset(a, 0, sizeof(a)); }
};

Mat5 mul5(const Mat5& A, const Mat5& B) {
    Mat5 C;
    for (int i = 0; i < 5; i++)
        for (int j = 0; j < 5; j++)
            for (int k = 0; k < 5; k++)
                C.a[i][j] = (C.a[i][j] + A.a[i][k] * B.a[k][j]) % MOD;
    return C;
}

Mat5 pow5(Mat5 M, ll n) {
    Mat5 R;
    for (int i = 0; i < 5; i++) R.a[i][i] = 1;
    while (n > 0) {
        if (n & 1) R = mul5(R, M);
        M = mul5(M, M);
        n >>= 1;
    }
    return R;
}

int main() {
    ll A, B, N;
    while (scanf("%lld %lld %lld", &A, &B, &N) == 3) {
        if (!A && !B && !N) break;

        ll fa, fa1, fb, fb1;
        fib(A, fa, fa1);
        fib(B, fb, fb1);

        ll p = fa * fb % MOD;
        ll q = fa1 * fb % MOD;
        ll r = fa * fb1 % MOD;
        ll s = fa1 * fb1 % MOD;
        ll S = p;

        if (N == 1) {
            printf("%lld\n", S);
            continue;
        }

        // Transition: p'=s, q'=r+s, r'=q+s, s'=p+q+r+s, S'=S+s
        Mat5 T;
        T.a[0][3] = 1;
        T.a[1][2] = 1; T.a[1][3] = 1;
        T.a[2][1] = 1; T.a[2][3] = 1;
        T.a[3][0] = 1; T.a[3][1] = 1; T.a[3][2] = 1; T.a[3][3] = 1;
        T.a[4][3] = 1; T.a[4][4] = 1;

        Mat5 TN = pow5(T, N - 1);

        ll vec[5] = {p, q, r, s, S};
        ll ans = 0;
        for (int j = 0; j < 5; j++)
            ans = (ans + TN.a[4][j] * vec[j]) % MOD;

        printf("%lld\n", ans);
    }

    return 0;
}

File: uri/estructuras/Fibos Sequence/1701.rs
use std::io::{self, Read, Write, BufWriter};

const MOD: u64 = 1_000_000_007;

type Mat2 = [[u64; 2]; 2];
type Mat5 = [[u64; 5]; 5];

fn mul2(a: &Mat2, b: &Mat2) -> Mat2 {
    let mut c = [[0u64; 2]; 2];
    for i in 0..2 {
        for j in 0..2 {
            for k in 0..2 {
                c[i][j] = (c[i][j] + a[i][k] * b[k][j]) % MOD;
            }
        }
    }
    c
}

fn pow2(mut m: Mat2, mut n: u64) -> Mat2 {
    let mut r: Mat2 = [[1, 0], [0, 1]];
    while n > 0 {
        if n & 1 == 1 { r = mul2(&r, &m); }
        m = mul2(&m, &m);
        n >>= 1;
    }
    r
}

fn fib(n: u64) -> (u64, u64) {
    if n == 0 { return (0, 1); }
    let base: Mat2 = [[1, 1], [1, 0]];
    let r = pow2(base, n);
    (r[1][0], r[0][0]) // (F(n), F(n+1))
}

fn mul5(a: &Mat5, b: &Mat5) -> Mat5 {
    let mut c = [[0u64; 5]; 5];
    for i in 0..5 {
        for j in 0..5 {
            for k in 0..5 {
                c[i][j] = (c[i][j] + a[i][k] * b[k][j]) % MOD;
            }
        }
    }
    c
}

fn pow5(mut m: Mat5, mut n: u64) -> Mat5 {
    let mut r = [[0u64; 5]; 5];
    for i in 0..5 { r[i][i] = 1; }
    while n > 0 {
        if n & 1 == 1 { r = mul5(&r, &m); }
        m = mul5(&m, &m);
        n >>= 1;
    }
    r
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());

    let mut tokens = input.split_whitespace();

    loop {
        let a: u64 = tokens.next().unwrap().parse().unwrap();
        let b: u64 = tokens.next().unwrap().parse().unwrap();
        let n: u64 = tokens.next().unwrap().parse().unwrap();

        if a == 0 && b == 0 && n == 0 { break; }

        let (fa, fa1) = fib(a);
        let (fb, fb1) = fib(b);

        let p = fa * fb % MOD;
        let q = fa1 * fb % MOD;
        let r = fa * fb1 % MOD;
        let s = fa1 * fb1 % MOD;
        let sum = p;

        if n == 1 {
            writeln!(out, "{}", sum).unwrap();
            continue;
        }

        // Transition: p'=s, q'=r+s, r'=q+s, s'=p+q+r+s, S'=S+s
        let mut t = [[0u64; 5]; 5];
        t[0][3] = 1;
        t[1][2] = 1; t[1][3] = 1;
        t[2][1] = 1; t[2][3] = 1;
        t[3][0] = 1; t[3][1] = 1; t[3][2] = 1; t[3][3] = 1;
        t[4][3] = 1; t[4][4] = 1;

        let tn = pow5(t, n - 1);

        let vec = [p, q, r, s, sum];
        let mut ans: u64 = 0;
        for j in 0..5 {
            ans = (ans + tn[4][j] * vec[j]) % MOD;
        }

        writeln!(out, "{}", ans).unwrap();
    }
}
